File: core/java/src/main/java/hara/lang/protocol/IWorkRef.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Immutable, transportable identity for one work run. */
@HaraProtocolBinding(
    namespace = "std.protocol.iworkref",
    name = "IWorkRef",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IWorkRef {
  @HaraMethod(value = "work-id", arity = 1)
  Object workId();
}

File: core/rust/src/lang/protocol/iworkref.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.iworkref",
    name = "IWorkRef",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IWorkRef {
    type Id;

    #[hara_method(value = "work-id", arity = 1)]
    fn work_id(&self) -> Self::Id;
}
